### ISSUE TITLE
distributed setup - unknown plugin 'flask'

### ISSUE THREAD

**Issue by sighn** [20c/vaping#91]
Sadly we can't get the distributed setup running:
```
Traceback (most recent call last):
  File "/usr/local/bin/vaping", line 8, in <module>
    sys.exit(cli())
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/click/decorators.py", line 73, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/vaping/cli.py", line 72, in start
    daemon = mk_daemon(ctx)
  File "/usr/local/lib/python3.6/site-packages/vaping/cli.py", line 47, in mk_daemon
    return vaping.daemon.Vaping(ctx.config)
  File "/usr/local/lib/python3.6/site-packages/vaping/daemon.py", line 82, in __init__
    plugin.instantiate(self.config['plugins'], self.plugin_context)
  File "/usr/local/lib/python3.6/site-packages/pluginmgr/config.py", line 114, in instantiate
    self.new_plugin(plugin_config, *args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/pluginmgr/config.py", line 64, in new_plugin
    obj = self._ctor(typ, config, *args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/pluginmgr/config.py", line 45, in _ctor
    return self.get_plugin_class(typ)(config, *args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/pluginmgr/__init__.py", line 202, in get_plugin_class
    raise ValueError("unknown plugin '%s'" % typ)
ValueError: unknown plugin 'flask'
```

We're using for the Master Server the example distributed config:
```
  - name: http
    type: flask
    host: 0.0.0.0
    port: 7021
    debug: true
    # set this to gunicorn or uwsgi depending on what you want to run with
    server: gunicorn
    async: gevent
    routes:
      /targets : graphsrv->targets
      /graph_data :
        methods:
          - POST
          - GET
        target: graphsrv->graph_data
      /graph : graphsrv->graph_view
      /overview_read_file : graphsrv->overview_read_file
      /: graphsrv->overview_view
```

```
python --version
Python 3.6.12

# pip install -U vaping vodka graphsrv flask
Requirement already up-to-date: vaping in /usr/local/lib/python3.6/site-packages (1.2.0)
Requirement already up-to-date: vodka in /usr/local/lib/python3.6/site-packages (2.3.0)
Requirement already up-to-date: graphsrv in /usr/local/lib/python3.6/site-packages (1.6.1)
Requirement already up-to-date: flask in /usr/local/lib/python3.6/site-packages (1.1.2)
Requirement already satisfied, skipping upgrade: gevent<3,>=1.1.2 in /usr/local/lib/python3.6/site-packages (from vaping) (1.5.0)
Requirement already satisfied, skipping upgrade: munge<1,>=0.6 in /usr/local/lib/python3.6/site-packages (from vaping) (0.6.0)
Requirement already satisfied, skipping upgrade: python-daemon<3,>=2.1.1 in /usr/local/lib/python3.6/site-packages (from vaping) (2.2.4)
Requirement already satisfied, skipping upgrade: click>=5.1 in /usr/local/lib/python3.6/site-packages (from vaping) (7.1.2)
Requirement already satisfied, skipping upgrade: pid<3,>=2.0.1 in /usr/local/lib/python3.6/site-packages (from vaping) (2.2.5)
Requirement already satisfied, skipping upgrade: pluginmgr<0.7,>=0.5.0 in /usr/local/lib/python3.6/site-packages (from vaping) (0.6.0)
Requirement already satisfied, skipping upgrade: future<1,>=0.14 in /usr/local/lib/python3.6/site-packages (from vodka) (0.18.2)
Requirement already satisfied, skipping upgrade: twentyc.tmpl==0.2.0 in /usr/local/lib/python3.6/site-packages (from vodka) (0.2.0)
Requirement already satisfied, skipping upgrade: Jinja2<3.0,>=2.8 in /usr/local/lib/python3.6/site-packages (from graphsrv) (2.11.2)
Requirement already satisfied, skipping upgrade: itsdangerous>=0.24 in /usr/local/lib/python3.6/site-packages (from flask) (1.1.0)
Requirement already satisfied, skipping upgrade: Werkzeug>=0.15 in /usr/local/lib/python3.6/site-packages (from flask) (1.0.1)
Requirement already satisfied, skipping upgrade: greenlet>=0.4.14; platform_python_implementation == "CPython" in /usr/local/lib/python3.6/site-packages (from gevent<3,>=1.1.2->vaping) (0.4.16)
Requirement already satisfied, skipping upgrade: requests<3,>=2.6.0 in /usr/local/lib/python3.6/site-packages (from munge<1,>=0.6->vaping) (2.24.0)
Requirement already satisfied, skipping upgrade: PyYAML>=3.10 in /usr/local/lib/python3.6/site-packages (from munge<1,>=0.6->vaping) (5.3.1)
Requirement already satisfied, skipping upgrade: setuptools in /usr/local/lib/python3.6/site-packages (from python-daemon<3,>=2.1.1->vaping) (49.6.0)
Requirement already satisfied, skipping upgrade: lockfile>=0.10 in /usr/local/lib/python3.6/site-packages (from python-daemon<3,>=2.1.1->vaping) (0.12.2)
Requirement already satisfied, skipping upgrade: docutils in /usr/local/lib/python3.6/site-packages (from python-daemon<3,>=2.1.1->vaping) (0.16)
Requirement already satisfied, skipping upgrade: MarkupSafe>=0.23 in /usr/local/lib/python3.6/site-packages (from Jinja2<3.0,>=2.8->graphsrv) (1.1.1)
Requirement already satisfied, skipping upgrade: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.6/site-packages (from requests<3,>=2.6.0->munge<1,>=0.6->vaping) (1.25.10)
Requirement already satisfied, skipping upgrade: chardet<4,>=3.0.2 in /usr/local/lib/python3.6/site-packages (from requests<3,>=2.6.0->munge<1,>=0.6->vaping) (3.0.4)
Requirement already satisfied, skipping upgrade: idna<3,>=2.5 in /usr/local/lib/python3.6/site-packages (from requests<3,>=2.6.0->munge<1,>=0.6->vaping) (2.10)
Requirement already satisfied, skipping upgrade: certifi>=2017.4.17 in /usr/local/lib/python3.6/site-packages (from requests<3,>=2.6.0->munge<1,>=0.6->vaping) (2020.6.20)
```

If you take a look in the plugins folder:
```
/vaping/plugins# ls -hal
total 76K
drwxr-xr-x 3 root root 4.0K Sep  1 10:22 .
drwxr-xr-x 4 root root 4.0K Sep  1 10:22 ..
-rw-r--r-- 1 root root  14K Sep  1 10:22 __init__.py
drwxr-xr-x 2 root root 4.0K Sep  1 10:22 __pycache__
-rw-r--r-- 1 root root 1.6K Sep  1 10:22 command.py
-rw-r--r-- 1 root root 4.7K Sep  1 10:22 fping.py
-rw-r--r-- 1 root root 3.5K Sep  1 10:22 fping_mtr.py
-rw-r--r-- 1 root root 2.0K Sep  1 10:22 graphite.py
-rw-r--r-- 1 root root  11K Sep  1 10:22 logparse.py
-rw-r--r-- 1 root root 1.4K Sep  1 10:22 rrd.py
-rw-r--r-- 1 root root 2.0K Sep  1 10:22 vodka.py
-rw-r--r-- 1 root root 1.6K Sep  1 10:22 whisper.py
-rw-r--r-- 1 root root 1.6K Sep  1 10:22 zeromq.py
```

Is the source missing the flask plugin or do we missing an depency?

**Comment by vegu** [comment]
Can you paste your vaping config please, also your output of `pip freeze`

**Comment by sighn** [comment]
Sure!
```
# pip freeze
certifi==2020.6.20
chardet==3.0.4
click==7.1.2
docutils==0.16
Flask==1.1.2
future==0.18.2
gevent==1.5.0
graphsrv==1.6.1
greenlet==0.4.16
idna==2.10
itsdangerous==1.1.0
Jinja2==2.11.2
lockfile==0.12.2
MarkupSafe==1.1.1
munge==0.6.0
pid==2.2.5
pluginmgr==0.6.0
python-daemon==2.2.4
PyYAML==5.3.1
pyzmq==19.0.2
requests==2.24.0
twentyc.tmpl==0.2.0
urllib3==1.25.10
vaping==1.2.0
vodka==2.3.0
Werkzeug==1.0.1
```

```
# cat config.yml

data:
  - type: fping
    handlers:
      - type: index
        index: host
      - type: store
        container: list
        limit: 500

apps:
  graphsrv:
    tmpl_engine: jinja2
    enabled: true
    # here is where we define host config
    groups:
      latency:
        public_dns:
          8.8.8.8:
            name: Google
            color: red
          1.1.1.1:
            name: Cloudflare
            color: blue

    graphs:
      multitarget:
        id_field: host
        type: multitarget
        plot_y: avg
        format_y: ms

      smokestack:
        id_field: host
        type: smokestack
        plot_y: avg

plugins:

  # zero mq probe plugin (latency name is important, so it can be
  # routed properly to the similarly named group)
  - name: latency
    type: zeromq
    data: fping
    interval: 1.0
    bind: tcp://0.0.0.0:6021
    async: thread

  - name: latency_2
    type: zeromq
    data_id: latency
    data: fping
    interval: 1.0
    bind: tcp://0.0.0.0:6022
    async: thread

  - name: latency_3
    type: zeromq
    data_id: latency
    data: fping
    interval: 1.0
    bind: tcp://0.0.0.0:6023
    async: thread

  - name: http
    type: flask
    host: 0.0.0.0
    port: 7021
    debug: true
    # set this to gunicorn or uwsgi depending on what you want to run with
    server: gunicorn
    async: gevent
    routes:
      /targets : graphsrv->targets
      /graph_data :
        methods:
          - POST
          - GET
        target: graphsrv->graph_data
      /graph : graphsrv->graph_view
      /overview_read_file : graphsrv->overview_read_file
      /: graphsrv->overview_view


logging:
  version: 1
  formatters:
    simple:
      format: '%(asctime)s - %(name)s - %(levelname)s: %(message)s'
  handlers:
    console:
      class: logging.StreamHandler
      level: DEBUG
      formatter: simple
      stream: ext://sys.stdout
  loggers:
    vodka:
      level: DEBUG
      handlers:
        - console
```

**Comment by vegu** [comment]
At a quick glance that all looks in order.

To answer your question the flask plugin lives with vodka, so you wouldn't find it in the vaping directory.

Since you are running the distributed setup, what is the command line you run to start up vodka (the web server)? 

From the traceback it looks like you're trying to start it through vaping which wont work for that setup.

Depending on which web server you are using it should be something like

```
export VODKA_HOME=examples/distributed_dns/vodka
gunicorn -b 0.0.0.0:7021 vodka.runners.wsgi:application
```

To start the web server process

**Comment by sighn** [comment]
Thank you this was indeed the error regarding the flask, but it seems that there is an error regarding the templates:

```
$ gunicorn -b 0.0.0.0:7021 vodka.runners.wsgi:application
[2020-09-01 16:56:15 +0200] [14418] [INFO] Starting gunicorn 20.0.4
[2020-09-01 16:56:15 +0200] [14418] [INFO] Listening at: http://0.0.0.0:7021 (14418)
[2020-09-01 16:56:15 +0200] [14418] [INFO] Using worker: sync
[2020-09-01 16:56:15 +0200] [14420] [INFO] Booting worker with pid: 14420
2020-09-01 16:56:16,017 - vodka - DEBUG: making sure configuration is sane ...
2020-09-01 16:56:16,020 - vodka - WARNING: [config unknown] 'apps.graphsrv.tmpl_engine is not a known configuration variable and has been ignored'
2020-09-01 16:56:16,031 - vodka - WARNING: [config deprecated] plugins.http.host is being deprecated in version Well be removed in 3.0, it's being replaced by the 'bind' config attribute
2020-09-01 16:56:16,032 - vodka - WARNING: [config deprecated] plugins.http.port is being deprecated in version Will be removed in 3.0, it's being replaced by the 'bind' config attribute
2020-09-01 16:56:16,032 - vodka - DEBUG: instantiating data types
2020-09-01 16:56:16,032 - vodka - DEBUG: instantiating plugins
[2020-09-01 16:56:16 +0200] [14418] [INFO] Shutting down: Master
[2020-09-01 16:56:16 +0200] [14418] [INFO] Reason: Worker failed to boot.
```

**Comment by vegu** [comment]
Don't think its the tmpl_engine warning, i think it's crashing on startup and gunicorn is eating the error, if i recall correctly there was an argument to pass to unicorn to show what the wsgi app is crashing on i dont recall it right now though.

You did set the VODKA_HOME env variable right?

**Comment by sighn** [comment]
Complete default
```
# git clone https://github.com/20c/vaping.git
Klone nach 'vaping' ...
remote: Enumerating objects: 49, done.
remote: Counting objects: 100% (49/49), done.
remote: Compressing objects: 100% (37/37), done.
remote: Total 1438 (delta 20), reused 32 (delta 9), pack-reused 1389
Empfange Objekte: 100% (1438/1438), 594.90 KiB | 1.83 MiB/s, Fertig.
Löse Unterschiede auf: 100% (868/868), Fertig.
# cd vaping/
~/vaping# export VODKA_HOME=examples/distributed_dns/vodka
~/vaping# gunicorn -b 0.0.0.0:7021 vodka.runners.wsgi:application
[2020-09-01 17:23:26 +0200] [31501] [INFO] Starting gunicorn 20.0.4
[2020-09-01 17:23:26 +0200] [31501] [INFO] Listening at: http://0.0.0.0:7021 (31501)
[2020-09-01 17:23:26 +0200] [31501] [INFO] Using worker: sync
[2020-09-01 17:23:26 +0200] [31504] [INFO] Booting worker with pid: 31504
2020-09-01 17:23:27,131 - vodka - DEBUG: making sure configuration is sane ...
2020-09-01 17:23:27,133 - vodka - WARNING: [config unknown] 'apps.graphsrv.tmpl_engine is not a known configuration variable and has been ignored'
2020-09-01 17:23:27,141 - vodka - WARNING: [config deprecated] plugins.http.host is being deprecated in version Well be removed in 3.0, it's being replaced by the 'bind' config attribute
2020-09-01 17:23:27,141 - vodka - WARNING: [config deprecated] plugins.http.port is being deprecated in version Will be removed in 3.0, it's being replaced by the 'bind' config attribute
2020-09-01 17:23:27,142 - vodka - DEBUG: instantiating data types
2020-09-01 17:23:27,142 - vodka - DEBUG: instantiating plugins
[2020-09-01 17:23:27 +0200] [31501] [INFO] Shutting down: Master
[2020-09-01 17:23:27 +0200] [31501] [INFO] Reason: Worker failed to boot.
```

**Comment by sighn** [comment]
Hope this additional output can help:
```
# gunicorn --log-level debug -b 0.0.0.0:7021 vodka.runners.wsgi:application
[2020-09-01 17:33:45 +0200] [31520] [DEBUG] Current configuration:
  config: None
  bind: ['0.0.0.0:7021']
  backlog: 2048
  workers: 1
  worker_class: sync
  threads: 1
  worker_connections: 1000
  max_requests: 0
  max_requests_jitter: 0
  timeout: 30
  graceful_timeout: 30
  keepalive: 2
  limit_request_line: 4094
  limit_request_fields: 100
  limit_request_field_size: 8190
  reload: False
  reload_engine: auto
  reload_extra_files: []
  spew: False
  check_config: False
  preload_app: False
  sendfile: None
  reuse_port: False
  chdir: /root/vaping
  daemon: False
  raw_env: []
  pidfile: None
  worker_tmp_dir: None
  user: 0
  group: 0
  umask: 0
  initgroups: False
  tmp_upload_dir: None
  secure_scheme_headers: {'X-FORWARDED-PROTOCOL': 'ssl', 'X-FORWARDED-PROTO': 'https', 'X-FORWARDED-SSL': 'on'}
  forwarded_allow_ips: ['127.0.0.1']
  accesslog: None
  disable_redirect_access_to_syslog: False
  access_log_format: %(h)s %(l)s %(u)s %(t)s "%(r)s" %(s)s %(b)s "%(f)s" "%(a)s"
  errorlog: -
  loglevel: debug
  capture_output: False
  logger_class: gunicorn.glogging.Logger
  logconfig: None
  logconfig_dict: {}
  syslog_addr: udp://localhost:514
  syslog: False
  syslog_prefix: None
  syslog_facility: user
  enable_stdio_inheritance: False
  statsd_host: None
  dogstatsd_tags:
  statsd_prefix:
  proc_name: None
  default_proc_name: vodka.runners.wsgi:application
  pythonpath: None
  paste: None
  on_starting: <function OnStarting.on_starting at 0x7ff6b02df7b8>
  on_reload: <function OnReload.on_reload at 0x7ff6b02df8c8>
  when_ready: <function WhenReady.when_ready at 0x7ff6b02df9d8>
  pre_fork: <function Prefork.pre_fork at 0x7ff6b02dfae8>
  post_fork: <function Postfork.post_fork at 0x7ff6b02dfbf8>
  post_worker_init: <function PostWorkerInit.post_worker_init at 0x7ff6b02dfd08>
  worker_int: <function WorkerInt.worker_int at 0x7ff6b02dfe18>
  worker_abort: <function WorkerAbort.worker_abort at 0x7ff6b02dff28>
  pre_exec: <function PreExec.pre_exec at 0x7ff6b02f80d0>
  pre_request: <function PreRequest.pre_request at 0x7ff6b02f81e0>
  post_request: <function PostRequest.post_request at 0x7ff6b02f8268>
  child_exit: <function ChildExit.child_exit at 0x7ff6b02f8378>
  worker_exit: <function WorkerExit.worker_exit at 0x7ff6b02f8488>
  nworkers_changed: <function NumWorkersChanged.nworkers_changed at 0x7ff6b02f8598>
  on_exit: <function OnExit.on_exit at 0x7ff6b02f86a8>
  proxy_protocol: False
  proxy_allow_ips: ['127.0.0.1']
  keyfile: None
  certfile: None
  ssl_version: 2
  cert_reqs: 0
  ca_certs: None
  suppress_ragged_eofs: True
  do_handshake_on_connect: False
  ciphers: None
  raw_paste_global_conf: []
  strip_header_spaces: False
[2020-09-01 17:33:45 +0200] [31520] [INFO] Starting gunicorn 20.0.4
[2020-09-01 17:33:45 +0200] [31520] [DEBUG] Arbiter booted
[2020-09-01 17:33:45 +0200] [31520] [INFO] Listening at: http://0.0.0.0:7021 (31520)
[2020-09-01 17:33:45 +0200] [31520] [INFO] Using worker: sync
[2020-09-01 17:33:45 +0200] [31523] [INFO] Booting worker with pid: 31523
[2020-09-01 17:33:45 +0200] [31520] [DEBUG] 1 workers
2020-09-01 17:33:45,921 - vodka - DEBUG: making sure configuration is sane ...
2020-09-01 17:33:45,923 - vodka - WARNING: [config unknown] 'apps.graphsrv.tmpl_engine is not a known configuration variable and has been ignored'
2020-09-01 17:33:45,933 - vodka - WARNING: [config deprecated] plugins.http.host is being deprecated in version Well be removed in 3.0, it's being replaced by the 'bind' config attribute
2020-09-01 17:33:45,933 - vodka - WARNING: [config deprecated] plugins.http.port is being deprecated in version Will be removed in 3.0, it's being replaced by the 'bind' config attribute
2020-09-01 17:33:45,934 - vodka - DEBUG: instantiating data types
2020-09-01 17:33:45,934 - vodka - DEBUG: instantiating plugins
[2020-09-01 17:33:45 +0200] [31520] [INFO] Shutting down: Master
[2020-09-01 17:33:45 +0200] [31520] [INFO] Reason: Worker failed to boot.
```

**Comment by vegu** [comment]
Odd, what happens if you pass the `--spew` argument to gunicorn?

**Comment by sighn** [comment]
A very long debug output - These are the last 100 lines before it ends:
```
# tail -n 100 output.log
logging:1521:         if (found == 0):
gunicorn.arbiter:346:         if self.pidfile is not None:
gunicorn.arbiter:348:         self.cfg.on_exit(self)
gunicorn.config:55:         if name not in self.settings:
gunicorn.config:57:         return self.settings[name].get()
gunicorn.config:299:         return self.value
gunicorn.config:1828:         pass
gunicorn.arbiter:349:         sys.exit(exit_status)
gunicorn.app.base:73:         except RuntimeError as e:
threading:1285:     tlock = _main_thread._tstate_lock
threading:1288:     assert tlock is not None
threading:1289:     assert tlock.locked()
threading:1290:     tlock.release()
threading:1291:     _main_thread._stop()
threading:983:         lock = self._tstate_lock
threading:984:         if lock is not None:
threading:985:             assert not lock.locked()
threading:986:         self._is_stopped = True
threading:987:         self._tstate_lock = None
threading:1292:     t = _pickSomeNonDaemonThread()
threading:1298:     for t in enumerate():
threading:1268:     with _active_limbo_lock:
threading:1269:         return list(_active.values()) + list(_limbo.values())
threading:1299:         if not t.daemon and t.is_alive():
threading:1133:         assert self._initialized, "Thread.__init__() not called"
threading:1134:         return self._daemonic
threading:1112:         assert self._initialized, "Thread.__init__() not called"
threading:1113:         if self._is_stopped or not self._started.is_set():
threading:1114:             return False
threading:1298:     for t in enumerate():
threading:1301:     return None
threading:1293:     while t:
logging:1937:     for wr in reversed(handlerList[:]):
logging:1940:         try:
logging:1941:             h = wr()
logging:1942:             if h:
logging:1943:                 try:
logging:1944:                     h.acquire()
logging:813:         if self.lock:
logging:814:             self.lock.acquire()
logging:1945:                     h.flush()
logging:975:         self.acquire()
logging:813:         if self.lock:
logging:814:             self.lock.acquire()
logging:976:         try:
logging:977:             if self.stream and hasattr(self.stream, "flush"):
logging:978:                 self.stream.flush()
logging:980:             self.release()
logging:820:         if self.lock:
logging:821:             self.lock.release()
logging:1946:                     h.close()
logging:895:         _acquireLock()
logging:225:     if _lock:
logging:226:         _lock.acquire()
logging:896:         try:    #unlikely to raise an exception, but you never know...
logging:897:             if self._name and self._name in _handlers:
logging:900:             _releaseLock()
logging:232:     if _lock:
logging:233:         _lock.release()
logging:1954:                     h.release()
logging:820:         if self.lock:
logging:821:             self.lock.release()
logging:1937:     for wr in reversed(handlerList[:]):
logging:1940:         try:
logging:1941:             h = wr()
logging:1942:             if h:
logging:1943:                 try:
logging:1944:                     h.acquire()
logging:813:         if self.lock:
logging:814:             self.lock.acquire()
logging:1945:                     h.flush()
logging:975:         self.acquire()
logging:813:         if self.lock:
logging:814:             self.lock.acquire()
logging:976:         try:
logging:977:             if self.stream and hasattr(self.stream, "flush"):
logging:1093:         return sys.stderr
logging:1093:         return sys.stderr
logging:978:                 self.stream.flush()
logging:1093:         return sys.stderr
logging:980:             self.release()
logging:820:         if self.lock:
logging:821:             self.lock.release()
logging:1946:                     h.close()
logging:895:         _acquireLock()
logging:225:     if _lock:
logging:226:         _lock.acquire()
logging:896:         try:    #unlikely to raise an exception, but you never know...
logging:897:             if self._name and self._name in _handlers:
logging:900:             _releaseLock()
logging:232:     if _lock:
logging:233:         _lock.release()
logging:1954:                     h.release()
logging:820:         if self.lock:
logging:821:             self.lock.release()
logging:1937:     for wr in reversed(handlerList[:]):
_weakrefset:39:             self = selfref()
_weakrefset:40:             if self is not None:
_weakrefset:41:                 if self._iterating:
_weakrefset:44:                     self.data.discard(item)
```

**Comment by vegu** [comment]
So i was able to reproduce in a fresh environment

And `--preload` is the argument i meant earlier that will show you the actual error, in my case it was missing install for zmq 

`pip install zmq`

Which is also missing from your pip install - if that doesn't fix it, please check your output of running gunicorn with --preload

**Comment by vegu** [comment]
Added a note about this to the docs in 1435138

**Comment by sighn** [comment]
Yes this fixed it - so it is not ```pyzmq==19.0.2``` and it should be zmq used instead.
Would be great if you made an note in the documentation for that.
Thank you for taking the time to debug this! Really appreciate it!

**Comment by vegu** [comment]
@sighn i think in py2.7 we had this working without having to install zmq alongside pyzmq and its something that was missed when adding py3 support

The weird part is i was now able to uninstall zmq and it still works afterwards.

In anycase having zmq installed doesn't seem to heard and be a requirement in some cases so i added a note to the docs for now.

Closing this ticket, thanks for the bug report / feedback :)
